circle-mlir/circle-mlir/lib/dialect/src/ConstFold.h

```diff
@@ -0,0 +1,263 @@
+    // in the N-dimension tensor. GetElementIndex returns
+    // the index in the flat representation of the original tensor
+    // to use.
+    const int64_t lhs_index = lhs_is_splat ? 0 : GetElementIndex(lhs_new_shape, current_index);
+    const int64_t rhs_index = rhs_is_splat ? 0 : GetElementIndex(rhs_new_shape, current_index);
+
+    new_values.push_back(
+      calculate(*(lhs_old_values.begin() + lhs_index), *(rhs_old_values.begin() + rhs_index)));
+    IncrementIndex(result_shape, &current_index);
+  }
+  return DenseElementsAttr::get(type, ArrayRef<ElementValueT>(new_values));
+}
+
+/// Performs const folding `calculate` with broadcast behavior on the two
+/// attributes `operand1` and `operand2` and returns the result if possible.
+/// This function assumes the two operands are verified to have value
+/// attributes of broadcastable types.
+template <class AttrElementT, class ElementValueT = typename AttrElementT::ValueType,
+          class CalculationT = llvm::function_ref<ElementValueT(ElementValueT, ElementValueT)>>
+Attribute ConstFoldBinaryOp(Type result_type, Attribute operand1, Attribute operand2,
+                            const CalculationT &calculate)
+{
+  if (operand1.dyn_cast_or_null<DenseElementsAttr>() &&
+      operand2.dyn_cast_or_null<DenseElementsAttr>())
+  {
+    return ConstFoldBinaryOpDenseDense<AttrElementT, ElementValueT>(
+      result_type, operand1.cast<DenseElementsAttr>(), operand2.cast<DenseElementsAttr>(),
+      calculate);
+  }
+
+  // TODO: support other attribute kinds
+
+  return {};
+}
+
+/// Performs const folding with broadcast behavior on the two attributes in
+/// `operands` and returns the result if possible.
+/// Depending on the given `resultType`, either `floatCalculate` or
+/// `intCalculate` is chosen to conduct the calculate.
+Attribute ConstFoldBinaryOp(Type result_type, ArrayRef<Attribute> operands,
+                            llvm::function_ref<APFloat(APFloat, APFloat)> float_calculate,
+                            llvm::function_ref<APInt(APInt, APInt)> int_calculate)
+{
+  // Note: All types are wrapped in tensor types in Circle. E.g., f32 is
+  // represented as tensor<f32>. So we are only handling tensor types here.
+  auto type = result_type.dyn_cast<ShapedType>();
+  if (!type)
+    return {};
+
+  auto elemType = type.getElementType();
+
+  if (elemType.isa<FloatType>())
+    return ConstFoldBinaryOp<FloatAttr>(result_type, operands[0], operands[1], float_calculate);
+
+  if (elemType.isSignlessInteger())
+    return ConstFoldBinaryOp<IntegerAttr>(result_type, operands[0], operands[1], int_calculate);
+
+  return {};
+}
+
+/// Performs const folding a attributes `operand` and returns the result if possible.
+/// The function currently asserts that the `result_type` to be a f32 tensor type.
+/// TODO: Extend this function to handle integral tensor for ops like "logical_not".
+Attribute ConstFoldUnaryOp(Type result_type, Attribute operand,
```

Comment:
It would be nice to add `inline` keyword to prevent multiple definition errors.
```suggestion
inline Attribute ConstFoldUnaryOp(Type result_type, Attribute operand,
```

Reply:
ditto